Localization test failures Return FormattableString for BoundTupleExpression.Display

MaStr11 opened the pull request:
### Customer scenario

Running unit test `OverloadResolutionTests.GenericInferenceOnInErrTuples` on non English machines failed due to localization issues. The root cause was that `BoundTupleExpression.Display` returned a string instead of a localizable string. The PR implements the solution proposed by @AlekseyTs in #24603. 
As noted in #24603 `BoundStackAllocArrayCreation.Display` a few lines down needs also to be fixed. This will be done once the fix `BoundTupleExpression.Display` is reviewed as being valid.

### Bugs this fixes

Fixes #24603 

### Workarounds, if any

Ignore test failures.

### Risk

The proposed solution requires to allow `FormattableString` types in `DiagnosticInfo.Arguments`. This might lead to unwanted code patterns in respect to DiagnosticInfo and localizability in the future.

### Performance impact

Low.

### Is this a regression from a previous update?

### Root cause analysis

Localization wasn't tested before.

### How was the bug found?

Contributor reported.

### Test documentation updated?

No.

MaStr11 commented:
@AlekseyTs could you review please?

AlekseyTs commented on src/Compilers/Core/Portable/Diagnostic/DiagnosticInfo.cs:
```diff
@@ -102,6 +102,11 @@ internal static void AssertMessageSerializable(object[] args)
                     continue;
                 }
 
+                if (arg is FormattableString)
```
>FormattableString [](start = 27, length = 17)

Perhaps make it more general by checking against IFormattable?

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/Formatting.cs:
```diff
@@ -82,17 +84,18 @@ public override object Display
 
-                return pooledBuilder.ToStringAndFree();
+                var format = pooledBuilder.ToStringAndFree();
+                var argumentDisplays = arguments.Select(a => a.Display).ToArray();
```
>var argumentDisplays = arguments.Select(a => a.Display).ToArray(); [](start = 16, length = 66)

Instead of doing this, allocate an array of the right size for arguments and populate it along with the format string.
```
var argumentsDisplay = new object[arguments.Length];
...
argumentsDisplay[i] = arguments[i].Display;
...
FormattableStringFactory.Create(format, argumentsDisplay)
```

AlekseyTs replied:
Make sure to remove ```using System.Linq;``` if it is no longer needed.

---
In reply to: [172283897](https://github.com/dotnet/roslyn/pull/25209#discussion_r172283897) [](ancestors = 172283897)

AlekseyTs commented:
Done with review pass (iteration 1)

MaStr11 commented:
@AlekseyTs 
I addressed your comments in cc0ea23. I also tried to fix `BoundStackAllocArrayCreation.Display` in bd62da1. The `Display` property is used in [StackAllocSpanExpressionsTests.ConditionalExpressionOnSpan_IncompatibleTypes](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Test/Semantic/Semantics/StackAllocSpanExpressionsTests.cs#L289) so the change I made is covered by a unit test.

AlekseyTs commented:
> StackAllocSpanExpressionsTests.ConditionalExpressionOnSpan_IncompatibleTypes so the change I made is covered by a unit test.

Was the test failing before your change? If not, we need to come up with a test that fails without the change.

MaStr11 commented:
No it didn't. It `Display` produces `stackalloc int[10]`, which isn't changed by the current culture (The other test also contained only language keywords but for some mysterious reasons `<null>` was translated by the German translators to `<NULL>`.) Can you give me some advice, how to write the test (were should it go and how should I test it).

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/Formatting.cs:
```diff
@@ -148,9 +150,7 @@ public override object Display
-        {
-            get { return string.Format(MessageID.IDS_StackAllocExpression.Localize().ToString(), ElementType, Count.Syntax); }
-        }
+            => ErrorFacts.GetMessageFormat(MessageID.IDS_StackAllocExpression, ElementType.ToDisplayString(), Count.Syntax.ToString());
```
>ToDisplayString [](start = 91, length = 15)

That builds string too early. Nothing should be formatted until client asks for the string

MaStr11 replied:
So we need a mixture of FormattableStringFactory.Create (delays the arguments) and ErrorFacts.GetMessageFormat (Delays the formatstring). I haven't seen anything that is able to do this. Can you give me a hint were to look?

AlekseyTs commented:
> The other test also contained only language keywords but for some mysterious reasons <null> was translated by the German translators to <NULL>.

What other test? This could be an indication of a premature localization.

MaStr11 commented:
> What other test?

The test that was failing on a German machine and made me fix BoundTupleExpression.Display. It was `OverloadResolutionTests.GenericInferenceOnInErrTuples` and is fixed by the changes until the last commit. The last commit addressed your comment in the original issue #24603.

> There is also an issue with BoundStackAllocArrayCreation.Display

Actually bd62da1 is a different concern. Sorry for being confusing here. I would also be fine with reverting bd62da1 and open a new issue for that.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/Formatting.cs:
```diff
@@ -148,9 +150,7 @@ public override object Display
     {
         public override object Display
-        {
-            get { return string.Format(MessageID.IDS_StackAllocExpression.Localize().ToString(), ElementType, Count.Syntax); }
```
>IDS_StackAllocExpression [](start = 49, length = 24)

It doesn't look like there is anything localizable in this resource - "stackalloc {0}[{1}]". I think we should remove it from resource file and hardcode the format string here. Then it becomes the same problem as for BoundTupleExpression. Note, it is fine to call ToString() on Count.Syntax and we should do this.

AlekseyTs commented:
Since the only part that can involve localization in BoundStackAllocArrayCreation is the type symbol, it would be extremely hard to come up with broken case. I think it will be fine to fix it without failing test

AlekseyTs commented:
@dotnet/roslyn-compiler, @jcouv, @VSadov Please review, need a second sign-off

AlekseyTs commented:
@dotnet/roslyn-compiler, @jcouv, @VSadov Please review, need a second sign-off

jcouv commented:
@jaredpar for ask-mode approval for 15.7. Thanks

jaredpar commented:
Approved

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Julien Couvreur <notifications@github.com>
Sent: Tuesday, March 6, 2018 1:47:06 PM
To: dotnet/roslyn
Cc: Jared Parsons; Mention
Subject: Re: [dotnet/roslyn] Localization test failures Return FormattableString for BoundTupleExpression.Display (#25209)


@jaredpar<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjaredpar&data=04%7C01%7Cjaredpar%40microsoft.com%7Ce616389c8bb846cd4fbe08d583abced1%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636559696338661473%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwifQ%3D%3D%7C-2&sdata=9hxhf18vCR9oD19veP1PMHtFMD4VHxPqMhNQE6XMygY%3D&reserved=0> for ask-mode approval for 15.7. Thanks

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Froslyn%2Fpull%2F25209%23issuecomment-370940038&data=04%7C01%7Cjaredpar%40microsoft.com%7Ce616389c8bb846cd4fbe08d583abced1%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636559696338661473%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwifQ%3D%3D%7C-2&sdata=WrRxsoM6ZnmGlaO%2Fa0Gh7o4bffsr95hsoq7DoMU0olI%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAAI-FxH06TUSwqo7hw9doSX9MGuZ7ZCnks5tbwPJgaJpZM4SbgHE&data=04%7C01%7Cjaredpar%40microsoft.com%7Ce616389c8bb846cd4fbe08d583abced1%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636559696338661473%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwifQ%3D%3D%7C-2&sdata=DF1sKFsBAE1CHwD08wxPgqGCK5ONqsF%2FjBcTMM6JYYs%3D&reserved=0>.

AlekseyTs commented:
@MaStr11 Thanks for the contribution!
